simapp/export.go

```diff
@@ -19,23 +18,31 @@ import (
 // file.
 func (app *SimApp) ExportAppStateAndValidators(
 	forZeroHeight bool, jailAllowedAddrs []string,
-) (appState json.RawMessage, validators []tmtypes.GenesisValidator, cp *abci.ConsensusParams, err error) {
-
+) (servertypes.ExportedApp, error) {
 	// as if they could withdraw from the start of the next block
 	ctx := app.NewContext(true, tmproto.Header{Height: app.LastBlockHeight()})
 
+	// We export at last height + 1, because that's the height at which
+	// Tendermint will start InitChain.
+	height := app.LastBlockHeight() + 1
```

Comment:
@alexanderbez I finally still decided to export at `N + 1`.

Tendermint will then call InitChain with `initial_height = N + 1` (or `1` if not set). So we InitChain(optionally also BeginBlock) at `N+1`, and the first commit on the new/upgraded chain will be at `N+1`.

Let me know if that makes sense to you.